src/data_source/storage/sql/client.rs

```diff
@@ -0,0 +1,582 @@
+        match Pin::new(&mut self.0).poll_read(cx, unfilled) {
+            Poll::Ready(Ok(bytes_read)) => {
+                // After the read completes, the first `bytes_read` of `unfilled` have now been
+                // filled. Increment the `filled` cursor within the `ReadBuf` to account for this.
+                buf.set_filled(filled + bytes_read);
+                Poll::Ready(Ok(()))
+            }
+            Poll::Ready(Err(err)) => Poll::Ready(Err(err)),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl tokio::io::AsyncWrite for TcpStream {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        Pin::new(&mut self.0).poll_write(cx, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        Pin::new(&mut self.0).poll_flush(cx)
+    }
+
+    fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        Pin::new(&mut self.0).poll_close(cx)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::super::testing::TmpDb;
+    use super::*;
+    use crate::testing::setup_test;
+    use futures::stream::TryStreamExt;
+
+    #[async_std::test]
+    async fn test_connection_closed_tx() {
+        setup_test();
+
+        let mut db = TmpDb::persistent().await;
+        let client = Client::new(db.config().client_config).await.unwrap();
+
+        let no_params: [i32; 0] = [];
+
+        tracing::info!("initialize database");
+        client
+            .batch_execute("CREATE TABLE test (key SERIAL PRIMARY KEY, value INTEGER)")
+            .await
+            .unwrap();
+        let modified = client
+            .execute_raw("INSERT INTO test (value) VALUES ($1), ($2)", [1, 2])
+            .await
+            .unwrap();
+        assert_eq!(modified, 2);
+
+        let get_values = || async {
+            let rows: Vec<_> = client
+                .query_raw("SELECT value FROM test", no_params)
+                .await
+                .unwrap()
+                .map_ok(|r| r.get::<_, i32>(0))
+                .try_collect()
+                .await
+                .unwrap();
+            rows
+        };
+        assert_eq!(get_values().await, [1, 2]);
+
+        // Start a transaction.
+        tracing::info!("begin transaction");
+        client.begin().await.unwrap();
+
+        // Initially, stuff works.
+        tracing::info!("query in transaction");
+        let modified = client
+            .execute_raw("INSERT INTO test (value) VALUES ($1), ($2)", [3, 4])
+            .await
+            .unwrap();
+        assert_eq!(modified, 2);
+        assert_eq!(get_values().await, [1, 2, 3, 4]);
+
+        // Shut down the database.
+        db.stop();
+
+        // Subsequent statements and queries within the transaction fail.
+        tracing::info!("query with DB shut down");
+        let err = client
```

Comment:
if this transaction fails then shouldn't the `in_progress` be set to `false`. After this, if I begin a new transaction then that should create a new connection or return an error if that times out?

Reply:
or maybe if I create a new transaction and the connection is closed then it retries ?

Reply:
Yes, but we haven't started a new transaction yet. We try that at the end of this test and make sure statements in the new transaction succeed